[bug] bazel client failure: Directory.getFilesList() is null

On latest buildfarm, the bazel client prints `Cannot invoke "build.bazel.remote.execution.v2.Directory.getFilesList()" because the return value of "java.util.Map.get(Object)" is null`.  Here is the full bazel client stacktrace:

```
java.io.IOException: com.google.devtools.build.lib.remote.ExecutionStatusException: UNKNOWN: Cannot invoke "build.bazel.remote.execution.v2.Directory.getFilesList()" because the return value of "java.util.Map.get(Object)" is null
	at com.google.devtools.build.lib.remote.GrpcRemoteExecutor.executeRemotely(GrpcRemoteExecutor.java:235)
	at com.google.devtools.build.lib.remote.RemoteExecutionService.executeRemotely(RemoteExecutionService.java:1258)
	at com.google.devtools.build.lib.remote.RemoteSpawnRunner.lambda$exec$2(RemoteSpawnRunner.java:268)
	at com.google.devtools.build.lib.remote.Retrier.execute(Retrier.java:244)
	at com.google.devtools.build.lib.remote.RemoteRetrier.execute(RemoteRetrier.java:125)
	at com.google.devtools.build.lib.remote.RemoteRetrier.execute(RemoteRetrier.java:114)
	at com.google.devtools.build.lib.remote.RemoteSpawnRunner.exec(RemoteSpawnRunner.java:243)
	at com.google.devtools.build.lib.exec.SpawnRunner.execAsync(SpawnRunner.java:245)
	at com.google.devtools.build.lib.exec.AbstractSpawnStrategy.exec(AbstractSpawnStrategy.java:146)
	at com.google.devtools.build.lib.exec.AbstractSpawnStrategy.exec(AbstractSpawnStrategy.java:108)
	at com.google.devtools.build.lib.actions.SpawnStrategy.beginExecution(SpawnStrategy.java:47)
	at com.google.devtools.build.lib.exec.SpawnStrategyResolver.beginExecution(SpawnStrategyResolver.java:68)
	at com.google.devtools.build.lib.exec.StandaloneTestStrategy.beginTestAttempt(StandaloneTestStrategy.java:440)
	at com.google.devtools.build.lib.exec.StandaloneTestStrategy.access$200(StandaloneTestStrategy.java:84)
	at com.google.devtools.build.lib.exec.StandaloneTestStrategy$StandaloneTestRunnerSpawn.beginExecution(StandaloneTestStrategy.java:672)
	at com.google.devtools.build.lib.analysis.test.TestRunnerAction.beginIfNotCancelled(TestRunnerAction.java:921)
	at com.google.devtools.build.lib.analysis.test.TestRunnerAction.beginExecution(TestRunnerAction.java:888)
	at com.google.devtools.build.lib.analysis.test.TestRunnerAction.execute(TestRunnerAction.java:946)
	at com.google.devtools.build.lib.analysis.test.TestRunnerAction.execute(TestRunnerAction.java:937)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$5.execute(SkyframeActionExecutor.java:907)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.continueAction(SkyframeActionExecutor.java:1076)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.run(SkyframeActionExecutor.java:1031)
	at com.google.devtools.build.lib.skyframe.ActionExecutionState.runStateMachine(ActionExecutionState.java:152)
	at com.google.devtools.build.lib.skyframe.ActionExecutionState.getResultOrDependOnFuture(ActionExecutionState.java:91)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor.executeAction(SkyframeActionExecutor.java:492)
	at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.checkCacheAndExecuteIfNeeded(ActionExecutionFunction.java:856)
	at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.computeInternal(ActionExecutionFunction.java:349)
	at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.compute(ActionExecutionFunction.java:169)
	at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:590)
	at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:382)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: com.google.devtools.build.lib.remote.ExecutionStatusException: UNKNOWN: Cannot invoke "build.bazel.remote.execution.v2.Directory.getFilesList()" because the return value of "java.util.Map.get(Object)" is null
	at com.google.devtools.build.lib.remote.GrpcRemoteExecutor.handleStatus(GrpcRemoteExecutor.java:71)
	at com.google.devtools.build.lib.remote.GrpcRemoteExecutor.getOperationResponse(GrpcRemoteExecutor.java:83)
	at com.google.devtools.build.lib.remote.GrpcRemoteExecutor.lambda$executeRemotely$2(GrpcRemoteExecutor.java:194)
	at com.google.devtools.build.lib.remote.Retrier.execute(Retrier.java:244)
	at com.google.devtools.build.lib.remote.RemoteRetrier.execute(RemoteRetrier.java:125)
	at com.google.devtools.build.lib.remote.RemoteRetrier.execute(RemoteRetrier.java:114)
	at com.google.devtools.build.lib.remote.GrpcRemoteExecutor.lambda$executeRemotely$3(GrpcRemoteExecutor.java:140)
	at com.google.devtools.build.lib.remote.util.Utils.refreshIfUnauthenticated(Utils.java:525)
	at com.google.devtools.build.lib.remote.GrpcRemoteExecutor.executeRemotely(GrpcRemoteExecutor.java:138)
	... 32 more
```

This message seems to be comming from buildfarm.  bazel would report it here:
https://github.com/bazelbuild/bazel/blob/a691e974d2e4c5fa4a469e1321b18d15ac7e9cfa/src/main/java/com/google/devtools/build/lib/remote/GrpcRemoteExecutor.java#L71

There are a few places in buildfarm where we call `Directory.getFilesList()`, but I'm trying to see why the directory would be null in the first place, and how this would be forwarded back to the client.

--- a/src/main/java/build/buildfarm/instance/server/AbstractServerInstance.java
+++ b/src/main/java/build/buildfarm/instance/server/AbstractServerInstance.java
@@ -739,7 +739,8 @@ public abstract class AbstractServerInstance implements Instance {
       Directory directory,
       Map<Digest, Directory> directoriesIndex,
       Consumer<String> onInputFile,
-      Consumer<String> onInputDirectory) {
+      Consumer<String> onInputDirectory,
+      PreconditionFailure.Builder preconditionFailure) {
     Stack<DirectoryNode> directoriesStack = new Stack<>();
     directoriesStack.addAll(directory.getDirectoriesList());
 
@@ -751,15 +752,29 @@ public abstract class AbstractServerInstance implements Instance {
           directoryPath.isEmpty() ? directoryName : (directoryPath + "/" + directoryName);
       onInputDirectory.accept(subDirectoryPath);
 
-      for (FileNode fileNode : directoriesIndex.get(directoryDigest).getFilesList()) {
-        String fileName = fileNode.getName();
-        String filePath = subDirectoryPath + "/" + fileName;
-        onInputFile.accept(filePath);
+      Directory subDirectory;
+      if (directoryDigest.getSizeBytes() == 0) {
+        subDirectory = Directory.getDefaultInstance();
+      } else {
+        subDirectory = directoriesIndex.get(directoryDigest);
       }
 
-      for (DirectoryNode subDirectoryNode :
-          directoriesIndex.get(directoryDigest).getDirectoriesList()) {
-        directoriesStack.push(subDirectoryNode);
+      if (subDirectory == null) {
+        preconditionFailure
+            .addViolationsBuilder()
+            .setType(VIOLATION_TYPE_MISSING)
+            .setSubject("blobs/" + DigestUtil.toString(directoryDigest))
+            .setDescription("The directory `/" + subDirectoryPath + "` was not found in the CAS.");
+      } else {
+        for (FileNode fileNode : subDirectory.getFilesList()) {
+          String fileName = fileNode.getName();
+          String filePath = subDirectoryPath + "/" + fileName;
+          onInputFile.accept(filePath);
+        }
+
+        for (DirectoryNode subDirectoryNode : subDirectory.getDirectoriesList()) {
+          directoriesStack.push(subDirectoryNode);
+        }
       }
     }
   }
@@ -881,7 +896,12 @@ public abstract class AbstractServerInstance implements Instance {
             subDirectory = directoriesIndex.get(directoryDigest);
           }
           enumerateActionInputDirectory(
-              subDirectoryPath, subDirectory, directoriesIndex, onInputFile, onInputDirectory);
+              subDirectoryPath,
+              subDirectory,
+              directoriesIndex,
+              onInputFile,
+              onInputDirectory,
+              preconditionFailure);
         } else {
           validateActionInputDirectoryDigest(
               subDirectoryPath,

--- a/src/test/java/build/buildfarm/instance/server/AbstractServerInstanceTest.java
+++ b/src/test/java/build/buildfarm/instance/server/AbstractServerInstanceTest.java
@@ -519,9 +519,10 @@ public class AbstractServerInstanceTest {
     assertThat(violation.getDescription()).isEqualTo("working directory is not an input directory");
   }
 
-  /**
-   * / -> valid dir bar/ -> missing dir with digest 'missing' and non-zero size foo/ -> missing dir
-   * with digest 'missing' and non-zero size
+  /*-
+   * / -> valid dir
+   *   bar/ -> missing dir with digest 'missing' and non-zero size
+   *   foo/ -> missing dir with digest 'missing' and non-zero size
    */
   @Test
   public void multipleIdenticalDirectoryMissingAreAllPreconditionFailures() {
@@ -562,6 +563,67 @@ public class AbstractServerInstanceTest {
     assertThat(violation.getType()).isEqualTo(VIOLATION_TYPE_MISSING);
     assertThat(violation.getSubject()).isEqualTo(missingSubject);
     assertThat(violation.getDescription()).isEqualTo(String.format(missingFmt, "foo"));
+  }
+
+  /*-
+   * / -> valid dir
+   *   bar/ -> valid dir
+   *     baz/ -> missing dir with digest 'missing-empty' and zero size
+   *     quux/ -> missing dir with digest 'missing' and non-zero size
+   *   foo/ -> valid dir with digest from /bar/, making it a copy of above
+   *
+   * Only duplicated-bar appears in the index
+   * Empty directory needs short circuit in all cases
+   * Result should be 2 missing directory paths, no errors
+   */
+  @Test
+  public void validationRevisitReplicatesPreconditionFailures() {
+    Digest missingEmptyDirectoryDigest = Digest.newBuilder().setHash("missing-empty").build();
+    Digest missingDirectoryDigest = Digest.newBuilder().setHash("missing").setSizeBytes(1).build();
+    Directory foo =
+        Directory.newBuilder()
+            .addAllDirectories(
+                ImmutableList.of(
+                    DirectoryNode.newBuilder()
+                        .setName("baz")
+                        .setDigest(missingEmptyDirectoryDigest)
+                        .build(),
+                    DirectoryNode.newBuilder()
+                        .setName("quux")
+                        .setDigest(missingDirectoryDigest)
+                        .build()))
+            .build();
+    Digest fooDigest = DIGEST_UTIL.compute(foo);
+    PreconditionFailure.Builder preconditionFailure = PreconditionFailure.newBuilder();
+    Directory root =
+        Directory.newBuilder()
+            .addAllDirectories(
+                ImmutableList.of(
+                    DirectoryNode.newBuilder().setName("bar").setDigest(fooDigest).build(),
+                    DirectoryNode.newBuilder().setName("foo").setDigest(fooDigest).build()))
+            .build();
+    AbstractServerInstance.validateActionInputDirectory(
+        ACTION_INPUT_ROOT_DIRECTORY_PATH,
+        root,
+        /* pathDigests=*/ new Stack<>(),
+        /* visited=*/ Sets.newHashSet(),
+        /* directoriesIndex=*/ ImmutableMap.of(fooDigest, foo),
+        /* onInputFiles=*/ file -> {},
+        /* onInputDirectories=*/ directory -> {},
+        /* onInputDigests=*/ digest -> {},
+        preconditionFailure);
+
+    String missingSubject = "blobs/" + DigestUtil.toString(missingDirectoryDigest);
+    String missingFmt = "The directory `/%s` was not found in the CAS.";
+    assertThat(preconditionFailure.getViolationsCount()).isEqualTo(2);
+    Violation violation = preconditionFailure.getViolationsList().get(0);
+    assertThat(violation.getType()).isEqualTo(VIOLATION_TYPE_MISSING);
+    assertThat(violation.getSubject()).isEqualTo(missingSubject);
+    assertThat(violation.getDescription()).isEqualTo(String.format(missingFmt, "bar/quux"));
+    violation = preconditionFailure.getViolationsList().get(1);
+    assertThat(violation.getType()).isEqualTo(VIOLATION_TYPE_MISSING);
+    assertThat(violation.getSubject()).isEqualTo(missingSubject);
+    assertThat(violation.getDescription()).isEqualTo(String.format(missingFmt, "foo/quux"));
   }
 
   @SuppressWarnings("unchecked")

